### DISCUSSION TITLE
Binary output problem

### DISCUSSION THREAD

**Discussion by windexplorer21** [OpenFAST/openfast#2430] [Q&A]
Hello everyone,

I’m running a steady wind speed step Monopile simulation, and I’m getting different results for the thrust coefficient when reading the .outb binary file (using ReadFASTbinary.m) compared to reading the .out text file (using ReadFASTtext.m).

Could anyone help me understand what might be causing this discrepancy?

Thank you in advance for your time and assistance!

Best regards,
windexplorer21

![image](https://github.com/user-attachments/assets/c4bed87e-583c-4d39-9c23-98afd8ff7277)

**Comment by jjonkman** [comment]
Dear @windexplorer21,

Did you by chance switch the labeling of your figures, with the one on the left referring to the _.out_ file and the one on the right referring to the _.outb_ file?  I would generally expect the binary file to have more precision, with the precision of the asci file determined by user-specified input `OutFmt` in the OpenFAST primary.  What have you set `OutFmt` to?

Best regards,

**Reply by windexplorer21** [reply]
Dear @jjonkman ,
Thank you for your answer. I'm confident about the labeling of the two figures. I have set the OutFmt in the .fst file to "ES10.3E2".

Best regards,
windexplorer21

**Comment by bjonkman** [comment]
@windexplorer21 , what value did you use for `OutFileFmt` in the primary OpenFAST input file?
```
3             OutFileFmt    - Format for tabular (time-marching) output file (switch)  {0: uncompressed binary [<RootName>.outb], 1: text file [<RootName>.out], 2: binary file [<RootName>.outb], 3: both 1 and 2, 5: Text and uncompressed binary files}

```
If you're using the "normal" compressed binary file, you might try the uncompressed file (e.g., option 5) to see if there is something strange in the binary compression algorithm.

**Reply by windexplorer21** [reply]
@bjonkman I used 3.

**Reply by bjonkman** [reply]
Can you try option 5 and see if that changes anything? The compression algorithm in the binary file is turned off with option 5.

**Reply by windexplorer21** [reply]
Dear @bjonkman,

I tried with OutFileFmt = 5, and it works. However I think it depends on the length of the simulation: for this test, I chose a Tmax of 3000 s. Just to be sure, I ran the simulation again with OutFileFmt = 3, and all four plots (OutFileFmt=3 .out, OutFileFmt=3 .outb, OutFileFmt=5 .out, and OutFileFmt=5 .outb) give me the same result.

I will try with my longer simulation and let you know.

Thank you!
Best regards,
@windexplorer21

**Reply by bjonkman** [reply]
The compression algorithm should care only about the _range_ of values in that particular channel, not the _length_ of the simulation. 

If there is an invalid number like **NaN** or **Inf** in the time series, that can cause an issue: you could search the .out file and see if there is something strange listed there (it would get read from the text file and then that point would not be plotted).

But actually, I think this may be an issue with the aero coefficients Ct, Cq, and Cp. We have modified the calculation of those channels slightly in the dev branch of OpenFAST (this change will be in v4.0.0). There can be instantaneous values of those coefficients that get large when the velocity in the denominator (v_dot_x) gets small... this can happen, for example, with a large change of wind direction. This causes the smaller values in those channels (in the range we actually care about) to lose precision in the binary file compression algorithm.

**Reply by windexplorer21** [reply]
Dear @bjonkman 
Thank you for your help. I checked for Nan values but I don't find any, I will update my OpenFAST and hope to solve this problem.

Best regards
Windexplorer21

**Reply by bjonkman** [reply]
I'm not sure the updated OpenFAST version will 100% solve the issue, but you can check. I'd probably first check the .out file to find the maximum value of that channel. If it's large, that would explain the issue. If not, there may be something else going on here.

**Comment by andrew-platt** [comment]
Which version of OpenFAST are you using?

**Reply by windexplorer21** [reply]
I'm using OpenFAST-v3.4.1
